R/dsb.R

```diff
@@ -0,0 +1,515 @@
+#'              ad_name,
+#'              ' value returned will be log transformed without background correction'),
+#'     )
+#'     ad = as.numeric(rep(x = 0, length(ad_name)))
+#'     names(ad) = ad_name
+#'     mu1 = c(mu1 , ad)
+#'     mu1 = mu1[match(rownames(norm_adt) , names(mu1) )]
+#'   }
+#'   norm_adt = apply( adt_log, 2, function(x) (x - mu1) )
+#'
+#'   #### STEP II
+#'   # dsb step II - calculate dsb technical component and regress out of ambient corrected values
+#'   if(isTRUE(denoise.counts)){
+#'     print(paste0('fitting models to each cell for dsb technical component and',
+#'                  ' removing cell to cell technical noise'))
+#'     cellwise_background_mean = apply(norm_adt, 2, function(x) {
+#'       g = mclust::Mclust(x, G=2, warn = FALSE, verbose = FALSE)
+#'       return(g$parameters$mean[1])
+#'     })
+#'     gc()
+#'     if (isTRUE(use.isotype.control)) {
+#'       noise_matrix = rbind(norm_adt[isotype.control.name.vec, ], cellwise_background_mean)
+#'       get_noise_vector = function(noise_matrix) {
+#'         g = stats::prcomp(t(noise_matrix), scale = TRUE)
+#'         return(g$x[ ,1])
+#'       }
+#'       # regress out eigenvector of noise matrix
+#'       noise_vector = get_noise_vector(noise_matrix)
+#'       norm_adt = limma::removeBatchEffect(norm_adt, covariates = noise_vector)
+#'     } else {
+#'       # regress out mu1
+#'       noise_vector = cellwise_background_mean
+#'       norm_adt = limma::removeBatchEffect(norm_adt, covariates = noise_vector)
+#'     }
+#'   }
+#'   # apply quantile clipping of outliers
+#'   if (isTRUE(quantile.clipping)) {
+#'     ql = apply(norm_adt, 1, FUN = stats::quantile, quantile.clip[1])
+#'     qh = apply(norm_adt, 1, FUN = stats::quantile, quantile.clip[2])
+#'     for (i in 1:nrow(norm_adt)) {
+#'       norm_adt[i, ] = ifelse(norm_adt[i, ] < ql[i], ql[i], norm_adt[i, ])
+#'       norm_adt[i, ] = ifelse(norm_adt[i, ] > qh[i], qh[i], norm_adt[i, ])
+#'     }
+#'   }
+#'   # write internal stats to R list to return if return.stats = TRUE
+#'   if(isTRUE(return.stats)) {
+#'     print('returning stats and results in a list; access normalized matrix with x$dsb_normalized_matrix')
+#'     protein_stats = list(
+#'
+#'       'raw cell matrix stats' = data.frame(
+#'         cell_mean = apply(adt_log, 1 , mean),
+#'         cell_sd = apply(adt_log, 1 , sd)),
+#'
+#'       'dsb normalized matrix stats' =
+#'         data.frame(
+#'           dsb_mean = apply(norm_adt, 1 , mean),
+#'           dsb_sd = apply(adt_log, 1 , sd))
+#'     )
+#'     if(isTRUE(denoise.counts)) {
+#'       technical_stats = cbind(t(noise_matrix), dsb_technical_component = noise_vector)
+#'     } else{
+#'       technical_stats = NULL
+#'     }
+#'     ret_obj = list(
+#'       'dsb_normalized_matrix' = norm_adt,
+#'       'technical_stats' = technical_stats,
+#'       'protein_stats' = protein_stats
+#'     )
+#'     return(ret_obj)
+#'   } else {
+#'     return(norm_adt)
+#'   }
+#' }
```

Comment:
I'm not sure what's going on here; it seems like all the code has been commented out except for an odd `source` call. All code in `R/internal.R` are available to other functions within SeuratWrappers so you shouldn't need the `source`

Reply:
Thanks for the note-for efficiency, in our workflow the 2 core functions are implemented just upstream of creation of a Seurat object using the dsb package from CRAN directly (this prevents Seurat object from having to access a potentially very large empty droplet matrix–see vignette). My thought was to have users load dsb package from CRAN directly then follow the seurat wrapper vignette which uses standard Seurat functions. This would also allow me to only maintain the CRAN pacage instead of having to push future patches to CRAN and SeuratWrappers. However, the functions are quite mature at this point. I'm happy to just include them in SeuratWrappers (the source call from before was just to make sure the user installed the package from CRAN). I will push code for functions directly.